backend/routes/users.js

```diff
@@ -0,0 +1,77 @@
+const express = require('express');
+const bcrypt = require('bcrypt');
+const { PrismaClient } = require('@prisma/client');
+const prisma = new PrismaClient();
+
+const router = express.Router();
+
+router.post('/', async (req, res) => {
+    const { username, email, password } = req.body;
+
+    try {
+        const existingUser = await prisma.user.findFirst({
+            where: {
+                OR: [
+                    { username: username },
+                    { email: email }
+                ]
+            }
+        });
+
+        if (existingUser) {
+            return res.status(400).json({ error: 'Username or email already exists' });
+        }
+
+        const hashedPassword = await bcrypt.hash(password, 10);
+
+        const newUser = await prisma.user.create({
+            data: { username, email, password: hashedPassword },
+        });
+
+        req.session.user = newUser;
+
+        res.json({ user: newUser });
+    } catch (error) {
+        console.error('Error during user sign up:', error);
+        res.status(500).json({ error: 'Server error' });
+    }
+});
+
+router.post('/login', async (req, res) => {
+    const { username, password } = req.body;
+
+    try {
+        const user = await prisma.user.findFirst({
+            where: { username }
+        });
+
+        if (!user) {
+            return res.status(401).json({ error: 'Invalid username or password' });
+        }
+
+        const isValidPassword = await bcrypt.compare(password, user.password);
+
+        if (!isValidPassword) {
+            return res.status(401).json({ error: 'Invalid username or password' });
+        }
+
+        req.session.user = user;
+
+        res.json({ user });
+    } catch (error) {
+        console.error('Error during user login:', error);
+        res.status(500).json({ error: 'Server error' });
+    }
+});
+
+router.post('/logout', (req, res) => {
+    req.session.destroy((err) => {
+        if (err) {
+            return res.status(500).json({ error: 'Failed to log out' });
+        }
+        res.clearCookie('connect.sid');
+        res.json({ message: 'Logged out successfully' });
+    });
+});
+
```

Comment:
may be out of scope, but wdyt about collecting user email in the signup? This will support "forgot password" workflow in the future (along with a plethora of other features, like email notifications when a new patch drops)

Reply:
the users submit an email when they signup. however, there's no restrictions to it yet. the user can submit emails that don't exist and there's no verification. i've thought about getting into it more once i've connected my comment section to my login (the login/signup should only appear if the user tries to comment/upvote without having an account). notifications are a part of the stretch goals, so i haven't done a lot of research on how to utilize emails yet.